### PR TITLE
aditional Updater Issue Selectors

### PR DESCRIPTION
**Pull request by jan-zajic**
Two aditional implementations of UpdaterIssueSelector next to DefaultUpdaterIssueSelector, that can be useful for other plugin developers or in workflow scripts. 
* select issues by jql query (JqlUpdaterIssueSelector)
* select issues explicitly by their keys (ExplicitUpdaterIssueSelector)

[Review on Reviewable](https://reviewable.io/reviews/jenkinsci/jira-plugin/82)

### CODE REVIEW

File: src/main/java/hudson/plugins/jira/updater/ExplicitUpdaterIssueSelector.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package hudson.plugins.jira.updater;
```
Comment (warden) [inline src/main/java/hudson/plugins/jira/updater/ExplicitUpdaterIssueSelector.java]: Move to hudson.plugins.jira.selector, along with the default one. Rename to AbstractIssueSelector.

File: src/main/java/hudson/plugins/jira/updater/ExplicitUpdaterIssueSelector.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.plugins.jira.JiraSite;
+import hudson.plugins.jira.UpdaterIssueSelector;
+
+public class ExplicitUpdaterIssueSelector extends UpdaterIssueSelector {
```
Comment (warden) [inline src/main/java/hudson/plugins/jira/updater/ExplicitUpdaterIssueSelector.java]: ExplicitUpdaterIssueSelector -> ExplicitIssueSelector/ExplicitStringIssueSelector/StringIssueSelector .

File: src/main/java/hudson/plugins/jira/updater/JqlUpdaterIssueSelector.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import com.atlassian.jira.rest.client.api.domain.Issue;
+import com.google.common.collect.Sets;
+
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.plugins.jira.JiraSession;
+import hudson.plugins.jira.JiraSite;
+import hudson.plugins.jira.UpdaterIssueSelector;
+
+public class JqlUpdaterIssueSelector extends UpdaterIssueSelector {
```
Comment (warden) [inline src/main/java/hudson/plugins/jira/updater/JqlUpdaterIssueSelector.java]: JqlUpdaterIssueSelector -> JqlIssueSelector

File: src/test/java/hudson/plugins/jira/updater/ExplicitIssueSelectorTest.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    @Test
+    public void testReturnsExplicitCollections() throws IOException {
+        JiraSession session = mock(JiraSession.class);
+        JiraSite site = mock(JiraSite.class);
+        when(site.getSession()).thenReturn(session);
+
+        ExplicitUpdaterIssueSelector jqlUpdaterIssueSelector = new ExplicitUpdaterIssueSelector(
+                Lists.newArrayList(TEST_KEY));
+        Set<String> findedIssueIds = jqlUpdaterIssueSelector.findIssueIds(null, site, null);
```
Comment (warden) [inline src/test/java/hudson/plugins/jira/updater/ExplicitIssueSelectorTest.java]: finded->found :)

File: src/test/java/hudson/plugins/jira/updater/JqlIssueSelectorTest.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    private final static String TEST_JQL = "key='EXAMPLE-1'";
+
+    @Test
+    public void testDontDependOnRunAndTaskListener() throws IOException {
+        JiraSession session = mock(JiraSession.class);
+        JiraSite site = mock(JiraSite.class);
+        when(site.getSession()).thenReturn(session);
+
+        JqlUpdaterIssueSelector jqlUpdaterIssueSelector = new JqlUpdaterIssueSelector(TEST_JQL);
+        Set<String> findedIssueIds = jqlUpdaterIssueSelector.findIssueIds(null, site, null);
```
Comment (warden) [inline src/test/java/hudson/plugins/jira/updater/JqlIssueSelectorTest.java]: .

File: src/test/java/hudson/plugins/jira/updater/JqlIssueSelectorTest.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+import hudson.plugins.jira.JiraSession;
+import hudson.plugins.jira.JiraSite;
+
+public class JqlIssueSelectorTest {
+
+    private final static String TEST_JQL = "key='EXAMPLE-1'";
+
+    @Test
+    public void testDontDependOnRunAndTaskListener() throws IOException {
+        JiraSession session = mock(JiraSession.class);
```
Comment (warden) [inline src/test/java/hudson/plugins/jira/updater/JqlIssueSelectorTest.java]: Create mocks in @Before setUp method.

**Comment by jan-zajic** [comment]
@warden, where you say "Move to hudson.plugins.jira.selector, along with the default one", you think 
move class DefaultUpdaterIssueSelector to package hudson.plugins.jira.selector? If so, it connotes 
some medium refactoring (because of the visibility of some methods). See my next commit what it means.
I think it's clearer and nice, but maybe some compatibility issues? Please decide, i revert it or squash
by your decision.

**Comment by warden** [comment]
Yes, that's what I meant. More than half of the commit are formatting changes, please rebase and re-commit with only the needed changes. From what I saw in this re-formatting mess, it looked good, but it is unreviewable currently...

**Comment by jan-zajic** [comment]
ok i fixed it

File: src/main/java/hudson/plugins/jira/selector/DefaultUpdaterIssueSelector.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+import hudson.plugins.jira.JiraCarryOverAction;
+import hudson.plugins.jira.JiraSite;
+import hudson.plugins.jira.Messages;
+import hudson.plugins.jira.RunScmChangeExtractor;
+import hudson.plugins.jira.UpdaterIssueSelector;
+import hudson.plugins.jira.listissuesparameter.JiraIssueParameterValue;
+import hudson.scm.ChangeLogSet;
+import hudson.scm.ChangeLogSet.Entry;
+
+public final class DefaultUpdaterIssueSelector extends UpdaterIssueSelector {
```
Comment (warden) [inline src/main/java/hudson/plugins/jira/selector/DefaultUpdaterIssueSelector.java]: DefaultIssueSelector

File: src/main/java/hudson/plugins/jira/selector/DefaultUpdaterIssueSelector.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+            for (AbstractBuild<?, ?> b : depc.getBuilds()) {
+                findIssues(b, ids, pattern, listener);
+            }
+        }
+        return ids;
+    }
+
+    /**
+     * @param pattern
+     *            pattern to use to match issue ids
```
Comment (warden) [inline src/main/java/hudson/plugins/jira/selector/DefaultUpdaterIssueSelector.java]: Line formatting/wrapping (should be in previous line)

File: src/test/java/hudson/plugins/jira/selector/DefaultUpdaterIssueSelectorTest.java [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+import hudson.model.FreeStyleBuild;
+import hudson.model.ParameterValue;
+import hudson.model.ParametersAction;
+import hudson.model.User;
+import hudson.plugins.jira.JiraSite;
+import hudson.plugins.jira.listissuesparameter.JiraIssueParameterValue;
+import hudson.scm.ChangeLogSet;
+import hudson.scm.ChangeLogSet.Entry;
+
+public class DefaultUpdaterIssueSelectorTest {
```
Comment (warden) [inline src/test/java/hudson/plugins/jira/selector/DefaultUpdaterIssueSelectorTest.java]: DefaultIssueSelectorTest

File: src/main/java/hudson/plugins/jira/selector/DefaultUpdaterIssueSelector.java (hunk 7 above)
Comment (warden) [inline src/main/java/hudson/plugins/jira/selector/DefaultUpdaterIssueSelector.java]: UpdaterIssueSelector -> AbstractIssueSelector and move to same package

**Pull request merged**